tests/pyroutils_test.py

```diff
@@ -25,3 +71,133 @@ def prior(num_samples=1):
+            estimator_dist.sample(torch.Size([3]))
+        x_samples = torch.distributions.Normal(theta, 1.0).sample(torch.Size([3]))
+
+    # Test log_prob method
+    assert estimator_dist.log_prob(x).shape == torch.Size([num_simulations])
+    assert estimator_dist.log_prob(x_samples).shape == torch.Size([3, num_simulations])
+
+    # Test expand method
+    estimator_dist_expanded = estimator_dist.expand(torch.Size([2, num_simulations]))
+    assert isinstance(estimator_dist_expanded, distribution_cls)
+    assert estimator_dist_expanded.batch_shape == torch.Size([2, num_simulations])
+    assert estimator_dist_expanded.event_shape == estimator_dist.event_shape
+    assert torch.equal(
+        estimator_dist_expanded.condition,
+        theta.expand(torch.Size([2, num_simulations, dim])),
+    )
+    assert estimator_dist_expanded.estimator == estimator_dist.estimator
+
+
+@pytest.mark.parametrize("num_dim", [3, 5])
+@pytest.mark.parametrize("num_trials", [1, 5])
+@pytest.mark.parametrize(
+    "trainer_cls, distribution_cls",
+    [
+        (NLE, ConditionalDensityEstimatorDistribution),
+        (NRE, RatioEstimatorDistribution),
+    ],
+)
+def test_pyro_gaussian_model(
+    trainer_cls,
+    distribution_cls,
+    num_dim,
+    num_trials,
+    num_simulations=500,
+    num_samples=500,
+    warmup_steps=500,
+):
+    """Test consistency of MCMC samples between the true and estimated likelihood."""
+    # Get data we will condition on
+    _, x_o = pyro_gaussian_model(num_trials=num_trials, num_dim=num_dim)
+
+    # Get MCMC samples from a model using the true likelihood
+    nuts_kernel = NUTS(pyro_gaussian_model)
+    mcmc = MCMC(nuts_kernel, num_samples=num_samples, warmup_steps=warmup_steps)
+    mcmc.run(x_o=x_o)
+    pyro_samples = torch.from_numpy(np.array(mcmc.get_samples()["theta"]))
+
+    # Simulated data from the true model and use it to train the estimator
+    prior = single_level_prior_theta(num_dim=num_dim)
+    theta = []
+    x = []
+    for _ in range(num_simulations):
+        ti, xi = pyro_gaussian_model(num_trials=1, num_dim=num_dim)
+        theta.append(torch.from_numpy(np.array(ti)))
+        x.append(torch.from_numpy(np.array(xi)))
+    theta = torch.stack(theta, dim=0).float()
+    x = torch.stack(x, dim=0).float().squeeze()
+    trainer = trainer_cls(prior=prior).append_simulations(theta=theta, x=x)
+    density_estimator = trainer.train()
+
+    # Define a model that uses the estimated likelihood
+    def sbi_pyro_model(likelihood, x_o=None):
+        theta = pyro.sample("theta", single_level_prior_theta(num_dim))  # (D,)
+
+        with pyro.plate("trials", num_trials):
+            x = pyro.sample("x", distribution_cls(likelihood, theta), obs=x_o)
+
+        if x_o is None:
+            return theta, x
+        else:
+            return x_o
+
+    if trainer_cls is not NRE:
+        theta, x = sbi_pyro_model(density_estimator)
+    sbi_pyro_model(density_estimator, x_o=x_o)
```

Comment:
not sure I understand these three lines. why the if-else for `NRE`, why the plain call to `sbi_pyro_model`?

Reply:
These calls are just to ensure that each branch of the Pyro model runs before we start sampling. The first line samples from the joint prior-prior-predictive distribution. But since `RatioEstimator` doesn't implement `sample`, that won't run here. The 2nd line just returns the conditioned observations without sampling the observations, so we run that for all.